### PR TITLE
Add support for gccs __builtin_mul_overflow

### PR DESCRIPTION
**Pull request by hannes-steffenhagen-diffblue**
Not the generic version.

This is a continuation of https://github.com/diffblue/cbmc/pull/5235. Once this is merged we’ll support all the non-generic overflow builtins.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: regression/cbmc/gcc_builtin_mul_overflow/main.c [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+#include <stdint.h>
+
+#ifdef __CPROVER
+#  define assume(X) __CPROVER_assume(X)
+#else
+#  define assume(X)                                                            \
+    do                                                                         \
+    {                                                                          \
+    } while(0)
+#endif
```
Comment (danpoe) [inline regression/cbmc/gcc_builtin_mul_overflow/main.c]: The `assume` macro doesn't seem to be used anywhere.

**Review by danpoe — APPROVED** [review]

File: regression/cbmc/gcc_builtin_mul_overflow/main.c (hunk 1 above)
Comment (hannes-steffenhagen-diffblue) [reply regression/cbmc/gcc_builtin_mul_overflow/main.c]: Good point

**Review by xbauch — APPROVED** [review]
Looks good.

File: regression/cbmc/gcc_builtin_mul_overflow/main.c [hunk 2]
```diff
@@ -0,0 +1,116 @@
+#include <assert.h>
+#include <limits.h>
+#include <stdint.h>
+
+#define A_VALUE_X_SO_THAT_X_TIMES_X_DOES_NOT_OVERFLOW(T)                       \
```
Comment (chrisr-diffblue) [inline regression/cbmc/gcc_builtin_mul_overflow/main.c:5]: That's quite a mouthful :-) Is `MAX_NON_OVERFLOWING_SQRT(T)` any better? Not sure, not a big deal, just floating a suggestion.

File: regression/cbmc/gcc_builtin_mul_overflow/main.c [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  (((T)(1)) << (sizeof(T) * CHAR_BIT) / 2 - 1)
```
Comment (chrisr-diffblue) [inline regression/cbmc/gcc_builtin_mul_overflow/main.c:6]: Is this correct for both signed and unsigned?

File: regression/cbmc/gcc_builtin_mul_overflow/main.c [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  int result;
+  assert(!__builtin_smul_overflow(1, 1, &result));
+  assert(result == 1);
+  int const lt_isqrt_of_int_max =
+    A_VALUE_X_SO_THAT_X_TIMES_X_DOES_NOT_OVERFLOW(int);
+  assert(!__builtin_smul_overflow(
+    lt_isqrt_of_int_max, lt_isqrt_of_int_max, &result));
+  assert(result == lt_isqrt_of_int_max * lt_isqrt_of_int_max);
+  assert(__builtin_smul_overflow(
+    lt_isqrt_of_int_max << 1, lt_isqrt_of_int_max << 1, &result));
```
Comment (chrisr-diffblue) [inline regression/cbmc/gcc_builtin_mul_overflow/main.c:19]: These would cause quite a large overflow - what about also testing with `lt_isqrt_of_int_max + 1` to get closer to a boundary condition? Ditto for the tests for the other types.

**Review by chrisr-diffblue — COMMENTED** [review]
Generally looks great - just a couple of queries/comments to consider.

File: regression/cbmc/gcc_builtin_mul_overflow/main.c (hunk 2 above)
Comment (hannes-steffenhagen-diffblue) [reply regression/cbmc/gcc_builtin_mul_overflow/main.c:5]: I went through several different iterations of this name. I didn’t particularly like any one of them, so I settled for the most “honest” one.

File: regression/cbmc/gcc_builtin_mul_overflow/main.c (hunk 3 above)
Comment (hannes-steffenhagen-diffblue) [reply regression/cbmc/gcc_builtin_mul_overflow/main.c:6]: Left shift for positive signed is defined, as is left shift for unsigned. Either way, this results in `(T)2^(n/2-1)`, and `2^(n/2-1)*2^(n/2-1) == 2^(n-2)` (`n` being the width of the type) which fits comfortably into both signed and unsigned (assuming twos complement which we are in CBMC anyway, and I think ones complement should still work too).

File: regression/cbmc/gcc_builtin_mul_overflow/main.c (hunk 4 above)
Comment (hannes-steffenhagen-diffblue) [reply regression/cbmc/gcc_builtin_mul_overflow/main.c:19]: @chrisr-diffblue `(lt_sqrt_of_int_max+1)*(lt_sqrt_of_int_max+1)` wouldn’t actually overflow at all with the way it’s defined.

**Review by chrisr-diffblue — APPROVED** [review]
